In [490]:
# Importing libraries
import pandas as pd
from os import listdir
from os.path import isfile, join
import xgboost as xgb
from datetime import datetime

# Reading the data
data_path = 'Data/'
log = pd.read_csv(data_path + 'LOG_WEB_201708.csv', sep=';')
clients = pd.read_csv(data_path + 'clients.csv', sep = ';')
cmd = pd.read_csv(data_path + 'CMD.csv', sep = ';')
cmd = cmd[cmd['ORDER_DATE'] >= '2017-08-01']
table = pd.read_csv(data_path + 'TABLE_CONVERSION.csv', sep = ';')

# Remove unamed column
log = log.loc[:, ~log.columns.str.contains('^Unnamed')]
clients = clients.loc[:, ~clients.columns.str.contains('^Unnamed')]
cmd = cmd.loc[:, ~cmd.columns.str.contains('^Unnamed')]
table = table.loc[:, ~table.columns.str.contains('^Unnamed')]

# Adjust data types - wrong types were causing issues
clients.CLIENT_NUMBER = [str(clients) for clients in clients.CLIENT_NUMBER]
cmd.CLIENT_NUMBER = [str(cmd) for cmd in cmd.CLIENT_NUMBER]
table.CLIENT_NUMBER = [str(table) for table in table.CLIENT_NUMBER]
table.VISITOR_ID = [str(table) for table in table.VISITOR_ID]
log.VISITOR_ID = [str(log) for log in log.VISITOR_ID]

# Cannot distinguish between multiple orders on the same day, so the table has been simplified.
cmd = cmd.groupby(['CLIENT_NUMBER', 'ORDER_DATE', 'CVIC']).sum()
# Convert back to typical DataFrame type
cmd = cmd.reset_index()

In [586]:
from multiprocessing import Pool

# Join cmd + table + log

In [506]:
cmd.CLIENT_NUMBER = [str(cmd) for cmd in cmd.CLIENT_NUMBER]
log['EVENT_SHORT'] = log.EVENT_DATE.str.slice(0,10)
data = log[log.VISITOR_ID != '0']
data = pd.merge(data, table, left_on='VISITOR_ID', right_on='VISITOR_ID', how='left')
data = pd.merge(data, cmd, left_on=['CLIENT_NUMBER', 'EVENT_SHORT'], right_on=['CLIENT_NUMBER', 'ORDER_DATE'], how='left')

In [621]:
data

VISITOR_ID CONNECTED_SESSION  ID_SESSION   SESSION_START_DATE  \
0          381225               OUI           1  2017-08-08 00:02:06   
1          563053               OUI          13  2017-08-08 00:01:32   
2          563053               OUI          13  2017-08-08 00:01:32   
3          706588               OUI          23  2017-08-08 00:01:29   
4          572702               OUI          64  2017-08-08 00:03:19   
5          572702               OUI          64  2017-08-08 00:03:19   
6          572702               OUI          64  2017-08-08 00:03:19   
7          698042               OUI         201  2017-08-08 00:14:41   
8          489527               OUI         203  2017-08-08 00:15:00   
9          489527               OUI         203  2017-08-08 00:15:00   
10         528544               OUI         232  2017-08-08 00:17:31   
11         706590               OUI         281  2017-08-08 00:22:35   
12         706590               OUI         281  2017-08-08 00:22:35   
13         484224               OUI         386  2017-08-08 00:33:41   
14         484224               OUI         386  2017-08-08 00:33:41   
15         484224               OUI         386  2017-08-08 00:33:41   
16         498172               OUI         444  2017-08-08 00:40:42   
17         706592               OUI         485  2017-08-08 00:45:40   
18         464818               OUI         517  2017-08-08 00:50:55   
19         464818               OUI         517  2017-08-08 00:50:55   
20         706591               OUI         536  2017-08-08 00:53:45   
21         706294               OUI         572  2017-08-08 01:01:15   
22         685876               OUI         589  2017-08-08 01:05:27   
23         685876               OUI         589  2017-08-08 01:05:27   
24         284469               OUI         611  2017-08-08 01:10:13   
25         284469               OUI         611  2017-08-08 01:10:13   
26         284469               OUI         611  2017-08-08 01:10:13   
27         623112               OUI         660  2017-08-08 01:22:12   
28         623112               OUI         660  2017-08-08 01:22:12   
29         623112               OUI         660  2017-08-08 01:22:12   
...           ...               ...         ...                  ...   
666704     632825               OUI       20287  2017-08-29 14:55:37   
666705     286176               OUI       20466  2017-08-29 15:00:20   
666706     286176               OUI       20466  2017-08-29 15:00:20   
666707     286176               OUI       20466  2017-08-29 15:00:20   
666708     286176               OUI       20466  2017-08-29 15:00:20   
666709     286176               OUI       20466  2017-08-29 15:00:20   
666710     711788               OUI       20705  2017-08-29 15:08:38   
666711     261647               OUI       23579  2017-08-29 16:47:27   
666712     295204               OUI       27180  2017-08-29 18:51:35   
666713     487207               OUI       27852  2017-08-29 19:20:31   
666714     296743               OUI        3950  2017-08-30 08:35:18   
666715     469453               OUI        8477  2017-08-30 09:27:52   
666716     469453               OUI        8477  2017-08-30 09:27:52   
666717     712067               OUI       13963  2017-08-30 11:06:54   
666718     712068               OUI       14326  2017-08-30 11:14:58   
666719     712073               OUI       15349  2017-08-30 11:38:55   
666720     340126               OUI       20535  2017-08-30 14:13:32   
666721     712213               OUI       23158  2017-08-30 14:42:31   
666722     163969               OUI       25495  2017-08-30 15:17:41   
666723     163969               OUI       25495  2017-08-30 15:17:41   
666724     688670               OUI       29342  2017-08-30 16:31:00   
666725     320981               OUI       43593  2017-08-30 23:05:20   
666726     691621               OUI        5839  2017-08-31 08:56:12   
666727     691621               OUI        5839

In [622]:
data.to_csv('data.csv', sep = ',')

# Visitor Journey

Looking at the logs of **August 2017** <br>
**Visitor 261647** first connects at **11:25 AM**
<br> <br>
Connected through **Sponsored Links**, on a **Computer** <br>
**19 minutes:** Address entered for delivery.  <br>
**22 minutes:** Accessed via phone to view account <br>
**28 minutes:** Puchase complete.<br>
**28 minutes:** Disconnected from website, doesn't reconnect this month
<br> <br>
**187.50 Euro Spent**

# What do I want to classify?
If a customer was convinced to purchase through magazine or online.

# Useful Features
- PAGES: Are they browsing the wares or going straight to the purchase?<br>
- GLOBAL_SOURCES: What caused them to connect to the website? <br>
- DEVICE_TYPE: Maybe magazine users use different devices <br>
- CVIC - Did they use a magazine coupon <br>
- PRE_TAX_AMOUNT <br>

# Feature Engineering

- time connected <br>
- number of unique pages visited
- T/F: Placed order 
- Time of Day

# Generating the Data Matrix
- VISITOR_ID <br>
- CLIENT_NUMBER <br>
- EVENT_SHORT <br>
- GLOBAL_SOURCES <br>
- DEVICE TYPE <br>
- CVIC <br>
- PRE_TAX_AMOUNT <br>


In [520]:
tmp = data[['VISITOR_ID', 'CLIENT_NUMBER', 'EVENT_SHORT', 'ID_SESSION', 'GLOBAL_SOURCES', 'DEVICE_TYPE', 'CVIC', 'PRE_TAX_AMOUNT', 'EVENT_DATE', 'PAGES']].copy()
tmp = tmp.fillna(0)
tmp['PURCHASE'] = True
tmp.loc[tmp.PRE_TAX_AMOUNT == 0, 'PURCHASE'] = False
tmp.columns = ['VISITOR_ID', 'CLIENT_NUMBER', 'DATE', 'ID_SESSION', 'SOURCE', 'DEVICE', 'PROMO', 'EXPENSE', 'EVENT_DATE', 'PAGES', 'PURCHASE']
tmp = tmp[['VISITOR_ID', 'CLIENT_NUMBER', 'DATE', 'ID_SESSION', 'SOURCE', 'DEVICE', 'PROMO', 'EXPENSE', 'PURCHASE', 'EVENT_DATE', 'PAGES']]
tmp['MONTH'] = tmp.EVENT_DATE.str.slice(0,7)
tmp = tmp[tmp.PURCHASE == True]

In [521]:
tmp.head()

VISITOR_ID         CLIENT_NUMBER        DATE  ID_SESSION        SOURCE  \
1     563053  -5859629033836166974  2017-08-08          13  Accès Direct   
2     563053  -5859629033836166974  2017-08-08          13  Accès Direct   
3     706588  -7174350147193854740  2017-08-08          23  Accès Direct   
4     572702    866042539231455965  2017-08-08          64       Moteurs   
5     572702    866042539231455965  2017-08-08          64       Moteurs   

       DEVICE  PROMO  EXPENSE  PURCHASE           EVENT_DATE    PAGES    MONTH  
1  Ordinateur  False   280.34      True  2017-08-08 00:03:27  accueil  2017-08  
2  Ordinateur  False   280.34      True  2017-08-08 00:22:04  accueil  2017-08  
3  Ordinateur  False    19.08      True  2017-08-08 00:01:32  accueil  2017-08  
4    Tablette   True    97.50      True  2017-08-08 00:03:19  accueil  2017-08  
5    Tablette   True    97.50      True  2017-08-08 00:09:42  accueil  2017-08

In [583]:
X = tmp.groupby(['CLIENT_NUMBER','ID_SESSION', 'MONTH', 'PROMO']).sum().reset_index()
X['SURF_TIME'] = 0
X.head()

CLIENT_NUMBER  ID_SESSION    MONTH  PROMO  EXPENSE  PURCHASE  \
0  -1008458086851481944       13945  2017-08  False  1813.00      14.0   
1  -1008458086851481944       18235  2017-08  False   388.50       3.0   
2  -1010661696429978923       37668  2017-08  False   403.75      19.0   
3  -1011583376281571893       12025  2017-08  False   632.00      16.0   
4  -1012644870962927739       30918  2017-08   True   215.01       3.0   

   SURF_TIME  
0          0  
1          0  
2          0  
3          0  
4          0

In [585]:
for client in X.CLIENT_NUMBER:
    for session in set(tmp[tmp.CLIENT_NUMBER == client].ID_SESSION):
        datetime.strptime(tmp[(tmp.ID_SESSION == session) & (tmp.CLIENT_NUMBER == client)].EVENT_DATE
, '%Y-%m-%d %H:%M:%S')
        b = tmp[(tmp.ID_SESSION == session) & (tmp.CLIENT_NUMBER == client)].iloc[-1].EVENT_DATE
        a = 
        b = datetime.strptime(b, '%Y-%m-%d %H:%M:%S')
        X[(X.CLIENT_NUMBER == client) & (X.ID_SESSION == session)].SURF_TIME = (b-a).total_seconds()/60

/anaconda3/lib/python3.6/site-packages/pandas/core/generic.py:3110: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


KeyboardInterrupt: 

In [619]:
def return_surftime(cn):
    session_surftime = []
    for session in set(tmp[tmp.CLIENT_NUMBER == cn].ID_SESSION):
        a = tmp[(tmp.ID_SESSION == session) & (tmp.CLIENT_NUMBER == cn)].iloc[0].EVENT_DATE
        b = tmp[(tmp.ID_SESSION == session) & (tmp.CLIENT_NUMBER == cn)].iloc[-1].EVENT_DATE
        a = datetime.strptime(a, '%Y-%m-%d %H:%M:%S')
        b = datetime.strptime(b, '%Y-%m-%d %H:%M:%S')
        session_surftime.append((b-a).total_seconds()/60)
    return(session_surftime)

In [620]:
p = Pool(2, maxtasksperchild=1)
surftime_res = p.map(return_surftime, X.CLIENT_NUMBER)
p.close()

ValueError: time data 'nan' does not match format '%Y-%m-%d %H:%M:%S'

In [596]:
surftime_res = [item for sublist in surftime_res for item in sublist]

In [618]:
a = str(tmp[(tmp.ID_SESSION == session) & (tmp.CLIENT_NUMBER == client)].EVENT_DATE.min())
a

'2017-08-18 23:42:40'

0:00:26


'2017-08-18 23:43:06'

In [564]:
datetime.strptime('2017-08-28 12:49:24', '%Y-%m-%d %H:%M:%S')

datetime.datetime(2017, 8, 28, 12, 49, 24)

# Pickle

In [487]:
with open(data_path+'client_by_date.pickle', 'rb') as handle:
   client_by_date = pickle.load(handle)

In [504]:
client_by_date

{'2017-08': [9089815182412087301,
  -4212914366395875314,
  4412715268887150606,
  7289012290777350169,
  870352525203243039,
  -4625196522819911645,
  -2031379451723022295,
  6441927626979868712,
  -5536008905043673038,
  1510620840604106801,
  6453316122314276914,
  7278064021490204723,
  7122309643670454325,
  5897545423665627191,
  -2506306593447968698,
  5397119728221356100,
  -7771995626234806199,
  -1597876238788263863,
  6241698716701524041,
  4210234185733210197,
  -4800809858005008293,
  -7885007154358452127,
  -3132438419034341280,
  5285474030847819876,
  8590983719385694308,
  6458856923129086055,
  -1507215425971453845,
  4559583669933178992,
  -4149535339675533,
  4659956769925496947,
  -3791399531358027651,
  6905247991116759170,
  811473898246832263,
  8976259379107758213,
  -2085207140827496307,
  9173330209602764939,
  1472052141509476498,
  4482980579790848145,
  -5541232508310421348,
  3464180607936069788,
  8189145422591557790,
  7877685402709917859,
  90650985451

# Testing

In [473]:
visit = pd.read_csv(data_path + 'VISITS_DATA_15h.csv')
visit = visit.loc[:, ~visit.columns.str.contains('^Unnamed')]
#visit = visit.fillna(0)
visit = visit[data.NB_ORDERS != 0]

visit.CLIENT_NUMBER = [str(data) for data in data.CLIENT_NUMBER]
visit.VISITOR_ID = [str(data) for data in data.VISITOR_ID]

/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2698: DtypeWarning: Columns (16,17) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [476]:
data.columns

Index(['VISITOR_ID', 'CLIENT_NUMBER', 'CREATION_DATE', 'NB_ORDERS',
       'TOT_AMOUNT_SPENT', 'FIRST_ORDER_DATE', 'LAST_ORDER_DATE',
       'ORDER_CHANNEL', 'CVIC', 'AVG_ORDER_AMOUNT', 'ACCOUNT_SENIORITY',
       'ORDER_SENIORITY', 'ORDER_LAST_ACTIVITY', 'ORDER_FREQUENCY',
       'TOTAL_VISITS', 'LAST_VISIT', 'FIRST_VISIT',
       'AVG_PAGES_VISITED_PER_VISIT', 'VISIT_SENIORITY', 'VISIT_LAST_ACTIVITY',
       'VISIT_FREQUENCY'],
      dtype='object')

In [482]:
#data.CLIENT_NUMBER.value_counts()
data[(data.CLIENT_NUMBER == '7721669007667066057')]

VISITOR_ID        CLIENT_NUMBER CREATION_DATE  NB_ORDERS  \
288416   469020.0  7721669007667066057    2015-03-11        1.0   
295059   476211.0  7721669007667066057    2015-03-11        1.0   
296442   477686.0  7721669007667066057    2015-03-11        1.0   
297777   479093.0  7721669007667066057    2015-03-11        1.0   
298296   479642.0  7721669007667066057    2015-03-11        1.0   
299020   480417.0  7721669007667066057    2015-03-11        1.0   
304579   486251.0  7721669007667066057    2015-03-11        1.0   
316586   499574.0  7721669007667066057    2015-03-11        1.0   
316602   499590.0  7721669007667066057    2015-03-11        1.0   

        TOT_AMOUNT_SPENT FIRST_ORDER_DATE LAST_ORDER_DATE ORDER_CHANNEL  \
288416            367.08       2015-05-05      2015-05-05      Internet   
295059            367.08       2015-05-05      2015-05-05      Internet   
296442            367.08       2015-05-05      2015-05-05      Internet   
297777            367.08       2015-05-05      2015-05-05      Internet   
298296            367.08       2015-05-05      2015-05-05      Internet   
299020            367.08       2015-05-05      2015-05-05      Internet   
304579            367.08       2015-05-05      2015-05-05      Internet   
316586            367.08       2015-05-05      2015-05-05      Internet   
316602            367.08       2015-05-05      2015-05-05      Internet   

         CVIC  AVG_ORDER_AMOUNT       ...         ORDER_SENIORITY  \
288416  False            367.08       ...                  1233.0   
295059  False            367.08       ...                  1233.0   
296442  False            367.08       ...                  1233.0   
297777  False            367.08       ...                  1233.0   
298296  False            367.08       ...                  1233.0   
299020  False            367.08       ...                  1233.0   
304579  False            367.08       ...                  1233.0   
316586  False            367.08       ...                  1233.0   
316602  False            367.08       ...                  1233.0   

        ORDER_LAST_ACTIVITY  ORDER_FREQUENCY  TOTAL_VISITS  LAST_VISIT  \
288416               1233.0              0.0           NaN         NaN   
295059               1233.0              0.0           NaN         NaN   
296442               1233.0              0.0           NaN         NaN   
297777               1233.0              0.0           NaN         NaN   
298296               1233.0              0.0           NaN         NaN   
299020               1233.0              0.0           NaN         NaN   
304579               1233.0              0.0           NaN         NaN   
316586               1233.0              0.0           NaN         NaN   
316602               1233.0              0.0           NaN         NaN   

       FIRST_VISIT AVG_PAGES_VISITED_PER_VISIT  VISIT_SENIORITY  \
288416         NaN                         NaN              NaN   
295059         NaN                         NaN              NaN   
296442         NaN                         NaN              NaN   
297777         NaN                         NaN              NaN   
298296         NaN                         NaN              NaN   
299020         NaN                         NaN              NaN   
304579         NaN                         NaN              NaN   
316586         NaN                         NaN              NaN   
316602         NaN                         NaN              NaN   

        VISIT_LAST_ACTIVITY  VISIT_FREQUENCY  
288416                  NaN              NaN  
295059                  NaN              NaN  
296442                  NaN              NaN  
297777                  NaN              NaN  
298296                  NaN              NaN  
299020                  NaN              NaN  
304579                  NaN              NaN  
316586                  NaN              NaN  
316602                  NaN              NaN  

[9 rows x 2

# VISITOR JOURNEY

In [230]:
v_17 = log[log.VISITOR_ID == '175752']
v_17.head()

VISITOR_ID CONNECTED_SESSION  ID_SESSION   SESSION_START_DATE  \
159645     175752               OUI       28914  2017-08-24 13:27:43   
159646     175752               OUI       28914  2017-08-24 13:27:43   
159647     175752               OUI       28914  2017-08-24 13:27:43   
182418     175752               OUI       14817  2017-08-27 10:17:25   
184863     175752               OUI       33484  2017-08-27 14:46:12   

                 EVENT_DATE    PAGES   GLOBAL_SOURCES DEVICE_TYPE  \
159645  2017-08-24 13:27:43  accueil  Email marketing  Ordinateur   
159646  2017-08-24 13:28:14  accueil  Email marketing  Ordinateur   
159647  2017-08-24 13:41:45  accueil  Email marketing  Ordinateur   
182418  2017-08-27 10:17:25  accueil     Accès Direct  Ordinateur   
184863  2017-08-27 14:46:12  accueil     Accès Direct  Ordinateur   

       DEVICE_MODEL  LOADINGS  
159645          NaN         1  
159646          NaN         1  
159647          NaN         1  
182418          NaN         1  
184863          NaN         1

In [235]:
v_17.sort_values(by=['EVENT_DATE'])

VISITOR_ID CONNECTED_SESSION  ID_SESSION   SESSION_START_DATE  \
159645      175752               OUI       28914  2017-08-24 13:27:43   
2091135     175752               OUI       28914  2017-08-24 13:27:43   
159646      175752               OUI       28914  2017-08-24 13:27:43   
278822      175752               OUI       28914  2017-08-24 13:27:43   
334433      175752               OUI       28914  2017-08-24 13:27:43   
667488      175752               OUI       28914  2017-08-24 13:27:43   
445710      175752               OUI       28914  2017-08-24 13:27:43   
445711      175752               OUI       28914  2017-08-24 13:27:43   
607990      175752               OUI       28914  2017-08-24 13:27:43   
547217      175752               OUI       28914  2017-08-24 13:27:43   
547218      175752               OUI       28914  2017-08-24 13:27:43   
565337      175752               OUI       28914  2017-08-24 13:27:43   
580843      175752               OUI       28914  2017-08-24 13:27:43   
594242      175752               OUI       28914  2017-08-24 13:27:43   
159647      175752               OUI       28914  2017-08-24 13:27:43   
182418      175752               OUI       14817  2017-08-27 10:17:25   
817674      175752               OUI       14817  2017-08-27 10:17:25   
524076      175752               OUI       14817  2017-08-27 10:17:25   
935047      175752               OUI       14817  2017-08-27 10:17:25   
1272059     175752               OUI       14817  2017-08-27 10:17:25   
346755      175752               OUI       14817  2017-08-27 10:17:25   
1272060     175752               OUI       14817  2017-08-27 10:17:25   
1331541     175752               OUI       14817  2017-08-27 10:17:25   
1272061     175752               OUI       14817  2017-08-27 10:17:25   
1193776     175752               OUI       14817  2017-08-27 10:17:25   
1272062     175752               OUI       14817  2017-08-27 10:17:25   
1616557     175752               OUI       14817  2017-08-27 10:17:25   
1272063     175752               OUI       14817  2017-08-27 10:17:25   
1498880     175752               OUI       14817  2017-08-27 10:17:25   
1272064     175752               OUI       14817  2017-08-27 10:17:25   
...            ...               ...         ...                  ...   
1181330     175752               OUI       12691  2017-08-29 10:55:54   
221826      175752               OUI       18234  2017-08-31 10:36:37   
822040      175752               OUI       18234  2017-08-31 10:36:37   
531102      175752               OUI       18234  2017-08-31 10:36:37   
841503      175752               OUI       18234  2017-08-31 10:36:37   
1276809     175752               OUI       18234  2017-08-31 10:36:37   
1616725     175752               OUI       18234  2017-08-31 10:36:37   
1276810     175752               OUI       18234  2017-08-31 10:36:37   
1617779     175752               OUI       18234  2017-08-31 10:36:37   
1276811     175752               OUI       18234  2017-08-31 10:36:37   
1616726     175752               OUI       18234  2017-08-31 10:36:37   
221827      175752               OUI       18234  2017-08-31 10:36:37   
225057      175752               OUI       33524  2017-08-31 14:57:38   
822382      175752               OUI       33524  2017-08-31 14:57:38   
531636      175752               OUI       33524  2017-08-31 14:57:38   
841523      175752               OUI       33524  2017-08-31 14:57:38   
691390      175752               OUI       33524  2017-08-31 14:57:38   
841524      175752               OUI       33524  2017-08-31 14:57:38   
691391      175752               OUI       33524  2017-08-31 14:57:38   
841525      175752               OUI       33524  2017-08-31 14:57:38   
489750      175752               OUI       33524  2017-08-31 14:57:38   
555107      175752               OUI       33524  2017-08-31 14:57:38   
555108      175752               OUI       33524  2017-08-31 14:57